File: google/binary-search/split-array-largest-sum.js
var splitArray = function (nums, m) {
    let [l, r] = [Math.max(...nums), nums.reduce((acc, val) => acc + val, 0)];
    let ans = r;

    function canSplit(largest) {
        let subarrayCount = 0;
        let sum = 0;

        for (let n of nums) {
            sum += n;
            if (sum > largest) {
                subarrayCount += 1;
                sum = n;
            }
        }
        return subarrayCount + 1 <= m;
    }

    while (l <= r) {
        const mid = Math.floor((l + r) / 2);
        if (canSplit(mid)) {
            ans = mid;
            r = mid - 1;
        } else l = mid + 1;
    }

    return ans;
};

console.log(splitArray([7, 2, 5, 10, 8], 2));

File: masteringDsAlgo/src/dsAlgo/dp/SplitArrayLargestSum.java
package dsAlgo.dp;

import java.util.Arrays;

public class SplitArrayLargestSum {
    private boolean canSplit(int largest, int[] arr, int m) {
        int sum = 0, subArrCnt = 0;
        for (int n: arr) {
            sum += n;
            if (sum > largest) {
                subArrCnt += 1;
                sum = n;
            }
        }
        return subArrCnt + 1 <= m;
    }

    public int splitArray(int[] arr, int m) {
        int l = Arrays.stream(arr).max().getAsInt();
        int r = Arrays.stream(arr).reduce(0, (a, b) -> a + b);

        int ans = r;
        while (l <= r) {
            int mid = (l + r) / 2;
            if (canSplit(mid, arr, m)) {
                ans = mid;
                r = mid - 1;
            } else l = mid + 1;
        }
        return ans;
    }
}
